### PR TITLE
Fixed #30386 -- Fixed primary key quoting in admin related object links.

### PR DESCRIPTION
**Pull request by Oluwayhemisi**
This is the branch I created to fix the issue of admin widgets not quoting primary keys. I fixed it from the server side and client side(creating a function in the Javascript file and implementing a quote function to quote the keys)  and i added test cases to check that the function i used to quote the primary keys for ticket-30386 works. I also used Javascript test and selenium to automate testing

### CODE REVIEW

**Comment by Oluwayhemisi** [comment]
@carltongibson Please can you check the implementation in the new commit I made? Java script encodeURIComponent() and encodeURI() didn't fix it. So I had to write a function, using the idea of encodeURIComponent() implementation to quote the primary key. I wrote a test for it. I also tested to update related object links when a value is passed, it quotes it and update the link.

**Review by carltongibson — COMMENTED** [review]
It's getting there. You've got the testing harnessing in the right place, but the escaping isn't right.

For Topping pk="_40": 
Change Form page URL: http://127.0.0.1:8000/admin/test_one/topping/_5F40/change/
Generated Pop-up URL: http://127.0.0.1:8000/admin/test_one/topping/%5F40/change/?_to_field=key&_popup=1 (without patch has just `_40`) 

So essentially the escaping is not the same. Look at the `quote` implementation in `admin.utils`. Match that and it would work: 

```
>>> from django.contrib.admin.utils import unquote
>>> unquote("%5F40")  # what you're sending now. 
'%5F40'
>>> unquote("_5F40")  # what's needed. 
'_40'. 
```

File: .gitignore [hunk 1]
```diff
@@ -16,3 +16,5 @@ tests/coverage_html/
 tests/.coverage*
 build/
 tests/report/
+requirements.txt
+/.idea
```
[the remark this answers is not in the document]
Comment (carltongibson) [reply .gitignore:20]: Yes, please. Remove this change. 👍 

> Please can you explain this "You can add these files to your local .git/info/exclude if you need to"?

Git ignores files listed there, as well as in the the `.gitignore` file. 

* [Ignore Files in Git Without Adding Them to .Gitignore | Luis Dalmolin](https://luisdalmolin.dev/blog/ignoring-files-in-git-without-gitignore/)
* [Git - Gitignore Documentation](https://git-scm.com/docs/gitignore)

File: tests/admin_widgets/tests.py [hunk 2]
```diff
@@ -1810,34 +1807,60 @@ def test_ForeignKey_using_to_field(self):
...
         )
-        house = House.objects.create(name="?a=b")
-        quoted_pk = quote(house.pk)
+        main_window = self.selenium.current_window_handle
 
-        field_name = self.selenium.find_element(By.ID, "id_house", quoted_pk).click()
-        field_name.send_keys(quoted_pk)
+        for house in self.house:
+            quoted_pk = quote(house.pk)
+            self.get_select_option('#id_house', quoted_pk).click()
```
Comment (carltongibson) [inline tests/admin_widgets/tests.py:1838]:
```
tests % ./runtests.py admin_widgets.tests.RelatedFieldWidgetSeleniumPrimaryKeyTests --selenium=firefox --parallel=1
Testing against Django installed in '/Users/carlton/Projects/Django/django/django'
Found 1 test(s).
Creating test database for alias 'default'...
System check identified no issues (1 silenced).
E
======================================================================
ERROR: test_quoted_primary_key (admin_widgets.tests.RelatedFieldWidgetSeleniumPrimaryKeyTests.test_quoted_primary_key)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/carlton/Projects/Django/django/tests/admin_widgets/tests.py", line 1838, in test_quoted_primary_key
    self.get_select_option('#id_house', quoted_pk).click()
    ^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'RelatedFieldWidgetSeleniumPrimaryKeyTests' object has no attribute 'get_select_option'

----------------------------------------------------------------------
Ran 1 test in 2.093s

FAILED (errors=1)
```

File: tests/admin_widgets/tests.py [hunk 3]
```diff
@@ -1810,34 +1807,60 @@ def test_ForeignKey_using_to_field(self):
...
+            self.wait_for_value('#id_house', quoted_pk)
 
-        save_button_css_selector = ".submit-row > input[type=submit]"
-        self.selenium.find_element(By.CSS_SELECTOR, save_button_css_selector.click)
+            field_name = self.selenium.find_element(By.ID, "id_house", ).click()
```
Comment (carltongibson) [inline tests/admin_widgets/tests.py:1841]: I don't think you want to `click` the element here.

**Comment by Oluwayhemisi** [comment]
@carltongibson please can you review the recent PR?

**Review by carltongibson — COMMENTED** [review]
Hi @Oluwayhemisi. 

You're hitting the error in the test at: 

```
        self.selenium.find_element(By.ID, "id_house").click()

selenium.common.exceptions.NoSuchElementException: Message: Unable to locate element: [id="id_house"]
```

In the previous line you hit save, which takes you back to the list view: 

```
        self.selenium.find_element(By.NAME, "_save").click()
```

If you run the single test case like so: 

```
% ./runtests.py --selenium=firefox --parallel=1 admin_widgets.tests.RelatedFieldWidgetSeleniumPrimaryKeyTests.test_create_houses_and_room --pdb
```

You'll enter the debugger at the error. (You have to go up `u` a few frames to get back to the test assertion.) 

You can interact with the browser window at that point though. 

If you go back to the room change form, and then click the pencil icon, it looks to work now. (But those are the steps to finish the test.) 

Make sense?

**Comment by Oluwayhemisi** [comment]
Yes it makes sense. My question is, does it mean i'm not suppose to click here                                                            
 self.selenium.find_element(By.ID, "id_house").click()
Since i already saved it here
 self.selenium.find_element(By.NAME, "_save").click()

**Comment by carltongibson** [comment]
@Oluwayhemisi You need to step through what's happening in the browser. Once you've created the Room you need to be on the change form page in order to check that the "Edit" pencil-icon popup has the correctly escaped URL. 

(So you either need to use the save-and-keep-editing button, or return to the change form from the change list… — or you could create the Room in code and then just load the change form once... — either way, what you're testing is that the PK value is correctly quoted and unquoted in the UI.)

**Comment by Oluwayhemisi** [comment]
> Yes it makes sense. My question is, does it mean i'm not suppose to click here self.selenium.find_element(By.ID, "id_house").click() Since i already saved it here self.selenium.find_element(By.NAME, "_save").click()

@Carlton, I have been able to fix it. But i ran into another error
When i create the Room it takes me to the change form page and I check that the "Edit" pencil-icon popup has the correct escaped URL.

But the issue is that I'm getting Timeout exception due to this line below 

`        self.wait_for_value("#id_name", house_name)

 `

If i remove the line of code above, on the browser, it takes me to the change form and when i try to check that the "Edit" pencil-icon popup has the correct escaped URL, the browser stops, though the URL shows it has been escaped but it doesn't display anything on the page

**Comment by Oluwayhemisi** [comment]
> > Yes it makes sense. My question is, does it mean i'm not suppose to click here self.selenium.find_element(By.ID, "id_house").click() Since i already saved it here self.selenium.find_element(By.NAME, "_save").click()
> 
> @carlton, I have been able to fix it. But i ran into another error When i create the Room it takes me to the change form page and I check that the "Edit" pencil-icon popup has the correct escaped URL.
> 
> But the issue is that I'm getting Timeout exception due to this line below
> 
> ` self.wait_for_value("#id_name", house_name)
> 
> `
> 
> If i remove the line of code above, on the browser, it takes me to the change form and when i try to check that the "Edit" pencil-icon popup has the correct escaped URL, the browser stops, though the URL shows it has been escaped but it doesn't display anything on the page

@carltongibson as regards to the issue of timeout exception, what i did is that i added time.sleep(5) to wait for the element to have the expected value. So it takes me to the page where the "Edit" pencil-icon popup would direct me when I click on it.  please what's do you think about it? am I on the right path?

**Comment by Oluwayhemisi** [comment]
@carltongibson please can you check the last PR i made?

**Comment by smithdc1** [comment]
Hey @Oluwayhemisi 👋

Carlton has moved on from his fellow role. 

If you're happy the review comments are addressed I suggest unticking the 'needs improvement' box on trac and I'm sure someone else will be along in due course to help review. 

https://www.djangoproject.com/weblog/2023/mar/31/welcome-our-new-fellow-natalia-bidart/

**Comment by Oluwayhemisi** [comment]
> Hey @Oluwayhemisi wave
> 
> Carlton has moved on from his fellow role.
> 
> If you're happy the review comments are addressed I suggest unticking the 'needs improvement' box on trac and I'm sure someone else will be along in due course to help review.
> 
> https://www.djangoproject.com/weblog/2023/mar/31/welcome-our-new-fellow-natalia-bidart/

Alright, thank you.

File: django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js [hunk 4]
```diff
@@ -185,6 +187,56 @@
...
+        const QUOTE_MAP = {};
+        const specialChars = '":/_#?;@&=+$,[]<>%\n\\';
+
+        for (let i = 0; i < specialChars.length; i++) {
+            const charCode = specialChars.charCodeAt(i);
+
+            // "This checks if the current character code is equal to 10 (which corresponds to the line feed character)";
+
+            if (charCode === 10) {
+                QUOTE_MAP[charCode] = '_0A';
```
Comment (sarahboyce) [inline django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js:201]:
```suggestion
            // Line feed \n character
            if (charCode === 10) {
                QUOTE_MAP[charCode] = '_0A';
```
I like the comment, think we can condense it a bit

File: django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js [hunk 5]
```diff
@@ -185,6 +187,56 @@
         win.close();
     }
 
+
+    function customEncodeURIComponent(s) {
```
Comment (sarahboyce) [inline django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js:191]:
```suggestion
    function customEncodeURIComponent(s) {
        // Ensure that primary key values do not confuse the admin URLs by escaping problematic characters.
        // This is the JavaScript equivalent to django.contrib.admin.utils.quote().
```
We might need a similar reference in `django.contrib.admin.utils.quote` or the python `QUOTE_MAP` about this function (so people know if they update something there, they should also update here)

File: django/contrib/admin/widgets.py [hunk 6]
```diff
@@ -196,7 +197,11 @@ def label_and_url_for_value(self, value):
                     obj._meta.app_label,
                     obj._meta.object_name.lower(),
                 ),
-                args=(obj.pk,),
+                args=(
+                    quote(
+                        obj.pk,
+                    ),
+                ),
```
Comment (sarahboyce) [inline django/contrib/admin/widgets.py:204]:
```suggestion
                args=(quote(obj.pk),),
```
I know black probably did this, I think the comma after `obj.pk` is not needed and it made black add all the extra lines

File: django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js [hunk 7]
```diff
@@ -185,6 +187,56 @@
...
+            const charToQuote = specialCharsList[i];
+            const quotedChar = customEncodeURIComponent(charToQuote);
+            quotedChars.push(quotedChar);
+        }
+        return quotedChars;
+    }
+
+
+
+    window.quoteSpecialChars = quoteSpecialChars;
```
Comment (sarahboyce) [inline django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js:236]: I think you have just defined this function to test `customEncodeURIComponent`, you should be able to remove this here and have these as example test cases

**Review by sarahboyce — COMMENTED** [review]
Hi @Oluwayhemisi :wave: this looks really good, thank you for your work on this :star: 

I'm not sure why the PR is a draft, I would change it so other people know it's ready for review :+1:

**Comment by sarahboyce** [comment]
buildbot, test on selenium.

**Comment by Oluwayhemisi** [comment]
> Hi @Oluwayhemisi wave this looks really good, thank you for your work on this star
> 
> I'm not sure why the PR is a draft, I would change it so other people know it's ready for review +1

@sarahboyce I will work on this

File: django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js (hunk 4 above)
Comment (Oluwayhemisi) [reply django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js:201]: > I like the comment, think we can condense it a bit
@sarahboyce thank you for the review, i will make the necessary modifications

File: django/contrib/admin/widgets.py [hunk 8]
```diff
@@ -257,6 +262,7 @@ def __init__(
         self.choices = widget.choices
         self.widget = widget
         self.rel = rel
+
```
Comment (felixxm) [inline django/contrib/admin/widgets.py:265]: Please revert all unrelated black lines and space changes.
```suggestion
```

File: django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js [hunk 9]
```diff
@@ -11,15 +11,15 @@
         relatedWindows.forEach(function(win) {
             if(!win.closed) {
                 win.dismissChildPopups();
-                win.close();    
+                win.close();
```
Comment (felixxm) [inline django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js:14]: Please revert all unrelated black lines and space changes.

File: django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js [hunk 10]
```diff
@@ -11,15 +11,15 @@
...
     }
 
     function setPopupIndex() {
         if(document.getElementsByName("_popup").length > 0) {
             const index = window.name.lastIndexOf("__") + 2;
-            popupIndex = parseInt(window.name.substring(index));   
+            popupIndex = parseInt(window.name.substring(index));
         } else {
             popupIndex = 0;
         }
```
Comment (felixxm) [inline django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js:25]: Please revert all unrelated black lines and space changes.

File: django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js [hunk 11]
```diff
@@ -76,9 +76,11 @@
...
         const value = $this.val();
         if (value) {
+
```
Comment (felixxm) [inline django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js:79]: Please revert all unrelated black lines and space changes.

**Review by felixxm — COMMENTED** [review]
@Oluwayhemisi Thanks for this patch :+1: Unfortunately, I've noticed that there is an issue when a key starts with `specialChars` (or maybe even contains one of them) , i.e. an instances edited in popups do not get back to the main object. For example:

https://user-images.githubusercontent.com/2865885/233315151-8cb261aa-b197-43eb-bbc4-6fdc2ddbac33.mp4

- topping for `Pizza (1)` starts with `_40` and doesn't update the topping in the main window.
- topping for `Pizza (6)` starts with `21` and updates the topping in the main window.

**Comment by nessita** [comment]
Hello @Oluwayhemisi! I will be trying to help you move this forward. With that in mind, I may push a few commits to this PR so we can continue iterating over it. Let me if you have any concerns!

**Comment by nessita** [comment]
@Oluwayhemisi I have made a few extra pushes to help resolving most comments in this PR. But there is still the issue reported by Mariusz, in that when a Topping is modified from the Pizzas change form (using the pencil action), when returning to the Pizza form the topping has not changed.

Do you think you could try to diagnose and fix that? Thank you!

**Comment by Oluwayhemisi** [comment]
@felixxm I will look into it. So sorry for the late response.

**Comment by Oluwayhemisi** [comment]
@nessita Thank you for the commits, I will go through it and give you a feedback. As regards to the issue reported by Mariusz, I will look into it and try and fix it.

File: .gitignore [hunk 12]
```diff
@@ -16,3 +16,6 @@ tests/coverage_html/
...
+venv
```
Comment (sarahboyce) [inline .gitignore:21]:
```suggestion
```
Let's revert these changes

File: tests/admin_views/tests.py
Comment (sarahboyce) [inline tests/admin_views/tests.py:1]: Can you revert all the changes in this file? It's just spacing and a print :+1:

**Review by sarahboyce — COMMENTED** [review]
Whoo hoo @Oluwayhemisi ! Amazing! Thank you for your updates!
I have some minor comments around cleaning up some spacing and such but I think this is ready for another review so I will update the ticket :+1:

**Review by felixxm — COMMENTED** [review]
I rebased, squashed, and tried to remove unrelated changes to make it reviewable.

File: django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js [hunk 13]
```diff
@@ -187,6 +190,74 @@
...
+        for (let i = 0; i < specialChars.length; i++) {
+            const charCode = specialChars.charCodeAt(i);
+            // Line feed \n character
+            if (charCode === 10) {
+                QUOTE_MAP[charCode] = '_0A';
+            } else {
+                const encodedChar = '_' + charCode.toString(16).toUpperCase();
+                QUOTE_MAP[charCode] = encodedChar;
+            }
+        }
```
Comment (felixxm) [inline django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js:209]: Can we have a static `QUOTE_MAP` instead of building it every time?

File: django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js [hunk 14]
```diff
@@ -187,6 +190,74 @@
...
+        for (let i = 0; i < specialChars.length; i++) {
+            const charCode = specialChars.charCodeAt(i);
+            // Line feed \n character
+            if (charCode === 10) {
+                UNQUOTE_MAP[charCode] = '_0A';
+            } else {
+                const encodedChar = '_' + charCode.toString(16).toUpperCase();
+                UNQUOTE_MAP[encodedChar] = String.fromCharCode(charCode);
+            }
+        }
```
Comment (felixxm) [inline django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js:237]: Can we have a static `UNQUOTE_MAP` instead of building it every time?

File: django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js [hunk 15]
```diff
@@ -187,6 +190,74 @@
...
+                const encodedChar = QUOTE_MAP[charCode] || s[i];
+                result += encodedChar;
+            }
+            return result;
+        } else {
+            return s;
+        }
+    }
+
+    function customDecodeURIComponent(s) {
```
Comment (felixxm) [inline django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js:224]: `custom` prefix in names are unnecessary.

File: django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js [hunk 16]
```diff
@@ -187,6 +190,74 @@
...
+                const encodedChar = '_' + charCode.toString(16).toUpperCase();
+                UNQUOTE_MAP[encodedChar] = String.fromCharCode(charCode);
+            }
+        }
+
+        if (typeof s === 'string') {
+            let result = '';
+            let i = 0;
+            while (i < s.length) {
+                if (s[i] === '_' && i < s.length - 2) {
```
Comment (felixxm) [inline django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js:243]: Can we `split()` `s` by `_` instead?

File: js_tests/admin/RelatedObjectLookups.test.js [hunk 17]
```diff
@@ -0,0 +1,55 @@
+/* global QUnit, RelatedObjectLookups */
+'use strict';
+QUnit.module('admin.RelatedObjectLookups');
+
+QUnit.test('quote value', function(assert) {
+    const result = customEncodeURIComponent('_40');
+    const expected = '_5F40';
+    assert.equal(result, expected, "expected _5F40");
+
```
Comment (felixxm) [inline js_tests/admin/RelatedObjectLookups.test.js:9]: Chop blank line.
```suggestion
```

File: tests/admin_widgets/tests.py [hunk 18]
```diff
@@ -762,6 +766,39 @@ def test_render_fk_as_pk_model(self):
...
+        w = widgets.ForeignKeyRawIdWidget(rel, widget_admin_site)
+
+        # apply quote function to primary key value
+        pk_quoted = quote(str(house.pk))
+
+        # render the widget
+        rendered = w.render("test", house.pk, attrs={})
+
+        # check that the primary key is properly quoted in the rendered HTML
+        self.assertIn(f'href="/admin_widgets/house/{pk_quoted}/change/"', rendered)
```
Comment (felixxm) [inline tests/admin_widgets/tests.py:800]: All comments are unnecessary as they don't add more explanation/
```suggestion
        pk_quoted = quote(str(house.pk))
        rendered = w.render("test", house.pk, attrs={})
        self.assertIn(f'href="/admin_widgets/house/{pk_quoted}/change/"', rendered)
```

File: tests/admin_widgets/tests.py [hunk 19]
```diff
@@ -1781,6 +1818,63 @@ def test_ForeignKey_using_to_field(self):
...
 
 
+class RelatedFieldWidgetSeleniumPrimaryKeyTests(AdminWidgetSeleniumTestCase):
+    def test_create_houses_and_room(self):
+        from selenium.webdriver.common.by import By
+        from selenium.webdriver.support.select import Select
+
+        self.admin_login(username="super", password="secret", login_url="/")
+
+        # Create a new House with a PK that needs quoting
```
Comment (felixxm) [inline tests/admin_widgets/tests.py:1828]:
```suggestion
        # Create a new House with a PK that needs quoting.
```

File: tests/admin_widgets/tests.py [hunk 20]
```diff
@@ -1781,6 +1818,63 @@ def test_ForeignKey_using_to_field(self):
...
+        house = House.objects.create(name="_40")
+        house_name = str(house.name)
+        self.selenium.get(
+            self.live_server_url + reverse("admin:admin_widgets_house_add")
+        )
+
+        self.selenium.find_element(By.ID, "id_name").send_keys(house_name)
+        self.selenium.find_element(By.NAME, "_continue").click()
+
+        # Check that the new House was created and listed in the change form
```
Comment (felixxm) [inline tests/admin_widgets/tests.py:1838]: To add more explanation, comments should state the expected behavior and omit prefixes like "Check that" since all tests check things. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

File: tests/admin_widgets/tests.py [hunk 21]
```diff
@@ -1781,6 +1818,63 @@ def test_ForeignKey_using_to_field(self):
...
+        self.selenium.find_element(By.ID, "id_name").send_keys(room_name)
+
+        self.selenium.find_element(By.NAME, "_continue").click()
+
+        save_button_css_selector = ".submit-row > input[type=submit]"
+
+        self.selenium.find_element(By.CSS_SELECTOR, save_button_css_selector)
+        self.selenium.find_element(By.ID, "view_id_house").click()
+        self.selenium.back()
+        time.sleep(60)
```
Comment (felixxm) [inline tests/admin_widgets/tests.py:1865]: This sleep is huuge :smile: Do we need it?

**Review by felixxm — COMMENTED** [review]
@Oluwayhemisi Thanks for updates :+1: I left more comments.

File: tests/admin_widgets/tests.py (hunk 21 above)
Comment (Oluwayhemisi) [reply tests/admin_widgets/tests.py:1865]: > This sleep is huuge 😄 Do we need it?
The sleep is not needed, I added it when i was trying to inspect the browser so the selenium test doesn't end. I thought I removed it but I will remove it. I will definitely work on the comments and blank spaces

Comment (Oluwayhemisi) [reply tests/admin_widgets/tests.py:1865]: The sleep is not needed, I added it when i was trying to inspect the browser so the selenium doesn't end. I thought i removed it but I will remove it.

File: tests/admin_widgets/tests.py (hunk 20 above)
Comment (Oluwayhemisi) [reply tests/admin_widgets/tests.py:1838]: Thank you, I will work on it.

**Comment by priyapahwa** [comment]
Hey, @Oluwayhemisi 
are you still working on this issue?

**Comment by felixxm** [comment]
Closing due to inactivity.
